Silabs EFM32 / EZR32 Happy Gecko Timer and PWM support #66893

viktorcsanak opened the pull request:
Added support for PWM to EFM32HG and `blinky_pwm` sample support for `efm32hg_slstk3400a`. The sample could not be tested on the `stk3400a` board since I don't have one at my disposal, but was working on a custom board featuring an `EZR32HG`. Since the `PF4 & 5` pins where the `UIF_LEDS` are connected according to below schematics can't be routed to Capture-Compare channels on any of the timers, it was set up on `PA0 - TIM0_CC0 #0` was used, which is available on the board headers.

https://www.silabs.com/documents/public/schematic-files/efm32hg-stk3400-schematics.pdf

An issue mentioned in  #66893 was that I could not set up PWM on TIM2 for a buzzer. Oddly enough it turns out that PWM will only work on any of the CC channels of any of the timers if all preceding timers has PWM enabled.

Commit:
dts: arm: silabs EFM32HG: Add timer and PWM support.

Add timer and PWM nodes to efm32hg.dtsi.
PWM nodes can't be used if previous PWM nodes aren't enabled.

Signed-off-by: Viktor Csanak <viktor.csanak@icbtech.rs>

Commit:
boards: arm: silabs efm32hg_slstk3400a: Add blinky_pwm sample support

Add aliases and nodes for sample support.

Signed-off-by: Viktor Csanak <viktor.csanak@icbtech.rs>

fkokosinski commented on boards/arm/efm32hg_slstk3400a/efm32hg_slstk3400a.dts:
```diff
@@ -101,3 +112,12 @@
+
+	pwm0: pwm {
+		status = "okay";
+		pin-location = <GECKO_LOCATION(0) GECKO_PORT_A GECKO_PIN(0)>;
```
Hi! Thanks for your contribution. Unfortunately, this way of setting up pinmuxing is considered deprecated in Zephyr. Please consult with "this page":https://docs.zephyrproject.org/latest/hardware/pinctrl/index.html to see how to use the Pinctrl API that is used in Zephyr for pinmuxing.

viktorcsanak replied:
Hi! I started working on the requested change. I pushed the progress I have so far, but the build is unstable for the time being. It fails while generating the updated configuration for the `pwm_gecko` driver, that should contain the `pinctrl_dev_config`.  I will get back to it when I can spare some time. I would appreciate if you could take a look at it in the meantime.

PS: Probably a ton of things will go wrong with the CI/CD checks, sorry for that in advance, I only published the changes so you can take a look at it.

Commit:
drivers: pinctrl_gecko: WIP: add support for PWM

Add initial support for TIMER CC pinmuxing.

Signed-off-by: Viktor Csanak <viktor.csanak@gmail.com>

Commit:
drivers: pwm_gecko: WIP: Add support for pinctrl.

Add initial support for pinmuxing with the pinctrl driver.
NOTE that the build is currently unstable (WIP).

Signed-off-by: Viktor Csanak <viktor.csanak@gmail.com>

Commit:
dts: arm: silabs EFM32HG: Add PWM pinctrl support.

Update PWM pinmultiplexing to use pinctrl.

Signed-off-by: Viktor Csanak <viktor.csanak@gmail.com>

Commit:
boards: arm: silabs efm32hg_slstk3400a: Add PWM pinctrl support.

Update PWM pinmultiplexing to use pinctrl.

Signed-off-by: Viktor Csanak <viktor.csanak@gmail.com>
